File: src/scenario/ui/_collapsibletable.js
/*
 * Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @file
 * @brief Expandable/collapsible tables.
 */


/** @var {object} `scenario.tables` package. */
scenario.tables = {};


// Add 'click' event listeners to `a.expand-all` and `a.collapse-all` buttons.
scenario.onLoad(() => {
    /**
     * @brief Configures `.expand-all` and `.collapse-all` buttons.
     * @param {"expand-all" | "collapse-all"} buttonClass Button class.
     * @param {"expanded" | "collapsed"} newState Final state wanted.
     * @returns {void}
     */
    function _configButton(buttonClass, newState) {
        for (/** @var {HTMLElement} */ const _button of scenario.findNodes(document.body, `a.button.${buttonClass}`)) {
            _button.addEventListener("click", (e) => {
                e.preventDefault();

                /** @var {string | null} */ const _tableId = scenario.getNamedObjectIdFromClasses(_button, "table");
                if (_tableId) {
                    scenario.tables._toggleAll(_tableId, newState);
                }
            });
        }
    }
    _configButton("expand-all", "expanded");
    _configButton("collapse-all", "collapsed");
});


// Add 'click' event listeners on `a.toggle-row` buttons.
scenario.onLoad(() => {
    for (/** @var {HTMLElement} */ const _mainRow of scenario.findNodes(document.body, "table.collapsible tr.main-row")) {
        for (/** @var {HTMLElement} */ const _button of scenario.findNodes(_mainRow, "a.button.toggle-row")) {
            _button.addEventListener("click", (e) => {
                e.preventDefault();

                scenario.tables._toggle(_mainRow);
            });
        }
    }
});


/**
 * @brief Expands or collapses all main rows.
 * @param {string} tableId Table identifier.
 * @param {"expanded" | "collapsed"} newState Final state wanted.
 * @returns {void}
 */
scenario.tables._toggleAll = (tableId, newState) => {
    console.debug(`scenario.tables._toggleAll('${tableId}', '${newState}')`);
    /** @var {number} */ let _count = 0;
    for (/** @var {HTMLElement} */ const _mainRow of scenario.findNodes(document.body, `table.collapsible.table=${tableId} tr.main-row`)) {
        scenario.tables._toggle(_mainRow, {newState: newState, collapsibleListItems: true});
        _count ++;
    }
    console.debug(`scenario.tables._toggleAll('${tableId}', '${newState}'): ${_count} main row(s) ${newState}`);
};


/**
 * @brief Expands or collapses `.collapsible-row` rows attached to a given `.main-row` row.
 * @param {HTMLElement} mainRow Main row to expand or collapse.
 * @param {"expanded" | "collapsed" | undefined} newState Final state wanted. Switch state if not provided.
 * @param {boolean | undefined} collapsibleListItems `true` to apply `newState` to related collapsible list items. Lists unchanged by default.
 * @returns {void}
 */
scenario.tables._toggle = (mainRow, {newState, collapsibleListItems} = {}) => {
    // Find 'main-row=...' id and 'expanded'/'collapsed' state from classes.
    /** @var {string | null} */ const _mainRowId = scenario.getNamedObjectIdFromClasses(mainRow, "main-row");
    if (! _mainRowId) {
        console.error(`No main row identifier found from ${mainRow} classes`);
        return;
    }
    /** @var {string | null} */ const _oldState = (
        mainRow.classList.contains("expanded") ? "expanded" :
        mainRow.classList.contains("collapsed") ? "collapsed" :
        null  // Neither 'expanded' nor 'collapsed'.
    );

    // Compute `newState` if not provided.
    if (! newState) {
        // Reverse status.
        switch (_oldState) {
            case "expanded": newState = "collapsed"; break;
            case "collapsed": newState = "expanded"; break;
        }
    }

    // Toggle from `_oldState` to `newState`.
    /** @var {HTMLElement | null} */ const _table = mainRow.parentElement;
    if (_table && _mainRowId && _oldState && newState) {
        console.debug(`Toggling ${_mainRowId}: ${_oldState} => ${newState}`);

        for (/** @var {HTMLElement} */ const _tr of scenario.findNodes(_table, `tr.main-row=${_mainRowId}`)) {
            // Remove old state.
            _tr.classList.remove("expanded");
            _tr.classList.remove("collapsed");

            // Set new state.
            _tr.classList.add(newState);

            // Adjust `tr.main-row a.button.toggle-row span` button text.
            if (_tr.classList.contains("main-row")) {
                for (/** @var {HTMLElement} */ const _buttonSpan of scenario.findNodes(_tr, "a.button.toggle-row span")) {
                    switch (newState) {
                        case "expanded": _buttonSpan.innerText = "-"; break;
                        case "collapsed": _buttonSpan.innerText = "+"; break;
                    }
                }
            }

            // Hide / show `tr.collapsible-row` rows.
            if (_tr.classList.contains("collapsible-row")) {
                switch (newState) {
                    case "expanded": _tr.style.visibility = "visible"; break;
                    case "collapsed": _tr.style.visibility = "collapse"; break;
                }
            }
        }
    }

    // Expand/collapse collapsible list items when required.
    if (collapsibleListItems && _table && _mainRowId) {
        for (/** @var {HTMLElement} */ const _tr of scenario.findNodes(_table, `tr.main-row=${_mainRowId}`)) {
            for (/** @var {HTMLElement} */ const _mainLi of scenario.findNodes(_tr, "li.collapsible.main-list-item")) {
                scenario.lists.toggle(_mainLi, {newState: newState});
            }
        }
    }
};

File: src/scenario/ui/_collapsibletable.py
# -*- coding: utf-8 -*-

# Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Expandable/collapsible tables.
"""

import typing

import scenario

if True:
    from ._collapsiblestate import CollapsibleState as _CollapsibleStateImpl  # @default-parameter-value
if typing.TYPE_CHECKING:
    from ._htmldoc import HtmlDocument as _HtmlDocumentType


class CollapsibleTableGenerator:
    """
    HTML generator for tables with collapsible rows.
    """

    def __init__(
            self,
            html,  # type: _HtmlDocumentType
            *,
            table_id,  # type: str
            default_state=_CollapsibleStateImpl.EXPANDED,  # type: _CollapsibleStateImpl
    ):  # type: (...) -> None
        """
        Instantiates a :class:`CollapsibleTableGenerator` with configurations.

        :param html: HTML output page to feed.
        :param table_id: Table identifier used in HTML classes.
        :param default_state: Default state for collapsible rows. Default is `expanded`.
        """
        #: HTML output page to feed.
        self.html = html  # type: _HtmlDocumentType
        #: Table identifier to set in HTML classes.
        self.table_id = table_id  # type: str
        #: Default state for collapsible rows.
        self.default_state = default_state

        #: Main row data, fed in :meth:`addmainrow()`.
        self._main_rows = []  # type: typing.List[CollapsibleTableGenerator._MainRow]

    def addexpandallbutton(self):  # type: (...) -> None
        """
        Adds a ``.expand-all`` button for the table being generated.
        """
        self.html.addnode(
            "a",
            classes=[
                "button", "expand-all",
                self.html.mknamedobjectidclass("table", self.table_id),
            ],
            href="#",
            text="Expand all",
        )

    def addcollapseallbutton(
            self,
    ):  # type: (...) -> None
        """
        Adds a ``.collapse-all`` button for the table being generated.
        """
        self.html.addnode(
            "a",
            classes=[
                "button", "collapse-all",
                self.html.mknamedobjectidclass("table", self.table_id),
            ],
            href="#",
            text="Collapse all",
        )

    def addtable(
            self,
            *,
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> _HtmlDocumentType.NodeContext
        """
        Starts generating the table.

        :param classes: Optional additional HTML classes.
        :return: HTML node context focused on the ``<table></table>`` created.
        """
        from ._htmldoc import HtmlDocument

        # Have '_collapsibletable.js' be embedded at the end of the HTML page.
        self.html.addfinaljs(scenario.Path(__file__).with_suffix(".js"))

        # Create the table node.
        _table_ctx = self.html.addnode(
            "table",
            classes=[
                *classes,
                "collapsible",
                self.html.mknamedobjectidclass("table", self.table_id),
            ],
        )  # type: _HtmlDocumentType.NodeContext

        # Return a `HtmlDocument.NodeContext` wrapper
        # to ensure `finalize()` be called for each main row at the end of the table definition.
        class _TableNodeContext(HtmlDocument.NodeContext):
            """
            :class:`._htmldoc.HtmlDocument.NodeContext` override
            in order to call :meth:`CollapsibleTableGenerator._MainRow.finalize()` for each main row
            at the end of the table definition.
            """

            def __init__(
                    self,
                    table_generator,  # type: CollapsibleTableGenerator
                    table_ctx,  # type: _HtmlDocumentType.NodeContext
            ):  # type: (...) -> None
                """
                :param table_generator: Current table generator with HTML output page to feed.
                :param table_ctx: Table node context.
                """
                HtmlDocument.NodeContext.__init__(
                    self,
                    html_doc=table_generator.html,
                    new_child=table_ctx.new_child,
                )

                #: Table generator being processed.
                self.table_generator = table_generator  # type: CollapsibleTableGenerator

            def __exit__(
                    self,
                    exc_type,  # type: typing.Any
                    exc_val,  # type: typing.Any
                    exc_tb,  # type: typing.Any
            ):  # type: (...) -> None
                super().__exit__(exc_type, exc_val, exc_tb)

                if not exc_type:
                    # Finalize table rows.
                    for _main_row in self.table_generator._main_rows:  # type: CollapsibleTableGenerator._MainRow  # noqa  ## Access to protected member
                        _main_row.finalize()

        return _TableNodeContext(
            table_generator=self,
            table_ctx=_table_ctx,
        )

    def addmainrow(
            self,
            *,
            main_row_id,  # type: str
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> _HtmlDocumentType.NodeContext
        """
        Adds a ``.main-row`` row in the table.

        :param main_row_id: Main row identifier used in HTML classes.
        :param classes: Optional additional HTML classes.
        :return: HTML node context focused on the ``<tr></tr>`` created.
        """
        self._main_rows.append(CollapsibleTableGenerator._MainRow(
            table_generator=self,
            id=main_row_id,
            tr_ctx=self.html.addnode(
                "tr",
                classes=[
                    *classes,
                    "main-row",
                    self.html.mknamedobjectidclass("table", self.table_id),
                    self.html.mknamedobjectidclass("main-row", main_row_id),
                ],
            ),
        ))
        return self._main_rows[-1].main_tr_ctx

    def addtogglebutton(self):  # type: (...) -> None
        """
        Adds a ``.toggle-row`` button (normally in a ``.main-row`` row).
        """
        self._main_rows[-1].toggle_button_a_ctx = self.html.addnode(
            "a",
            classes=[
                "button", "toggle-row",
                self.html.mknamedobjectidclass("table", self.table_id),
                self.html.mknamedobjectidclass("main-row", self._main_rows[-1].id),
            ],
        )
        with self._main_rows[-1].toggle_button_a_ctx:
            # Create a span node without text.
            # The text will be set when the table is terminated.
            self._main_rows[-1].toggle_button_span_ctx = self.html.addnode("span")

    def addcollapsiblerow(
            self,
            *,
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> _HtmlDocumentType.NodeContext
        """
        Adds a ``.collapsible-row`` row in the table.

        :param classes: Optional additional HTML classes.
        :return: HTML node context focused on the ``<tr></tr>`` created.
        """
        self._main_rows[-1].collapsible_tr_ctxs.append(
            self.html.addnode(
                "tr",
                classes=[
                    *classes,
                    "collapsible-row",
                    self.html.mknamedobjectidclass("table", self.table_id),
                    self.html.mknamedobjectidclass("main-row", self._main_rows[-1].id),
                ],
            )
        )
        return self._main_rows[-1].collapsible_tr_ctxs[-1]

    class _MainRow:
        """
        Main row information.
        """

        def __init__(
                self,
                table_generator,  # type: CollapsibleTableGenerator
                id,  # type: str  # noqa  ## Shadows built-in name 'id'
                tr_ctx,  # type: _HtmlDocumentType.NodeContext
        ):  # type: (...) -> None
            """
            :param table_generator: Owner table generator.
            :param id: Main row identifier.
            :param tr_ctx: `tr.main-row` node context.
            """
            #: Owner table generator.
            self.table_generator = table_generator
            #: Main row identifier.
            self.id = id  # type: str
            #: Main row ``<tr></tr>`` node context.
            self.main_tr_ctx = tr_ctx  # type: _HtmlDocumentType.NodeContext
            #: ``.toggle-row`` button ``<a></a>`` node context, saved in :meth:`CollapsibleTableGenerator.addtogglebutton()`.
            self.toggle_button_a_ctx = None  # type: typing.Optional[_HtmlDocumentType.NodeContext]
            #: ``.toggle-row`` button ``<span></span>`` node context, saved in :meth:`CollapsibleTableGenerator.addtogglebutton()`.
            self.toggle_button_span_ctx = None  # type: typing.Optional[_HtmlDocumentType.NodeContext]
            #: ``<tr></tr>`` node contexts of related collapsible rows, fed in :meth:`CollapsibleTableGenerator.addcollapsiblerow()`.
            self.collapsible_tr_ctxs = []  # type: typing.List[_HtmlDocumentType.NodeContext]

        @property
        def html(self):  # type: () -> _HtmlDocumentType
            """
            Shortcut to HTML output page to feed.
            """
            return self.table_generator.html

        def finalize(self):  # type: (...) -> None
            """
            Finalizes the main collapsible row.

            Depending on whether on the main row has related collapsible rows attached to it or not:

            - sets `.expanded`/`.collapsed` state classes,
            - sets `.toggle-row` button text,
            - sets related collapsible rows visibility.
            """
            if not self.collapsible_tr_ctxs:
                # No `.collapsible-row`s attached.

                # Toggle button text.
                if self.toggle_button_span_ctx is not None:
                    with self.toggle_button_span_ctx:
                        self.html.addtext("o")

            else:
                # `.collapsible-row`s attached.

                # Main row `.expanded`/`.collapsed` class.
                with self.main_tr_ctx:
                    self.html.addclass(self.table_generator.default_state)

                # Toggle button `.expanded`/`.collapsed` class.
                if self.toggle_button_a_ctx is not None:
                    with self.toggle_button_a_ctx:
                        self.html.addclass(self.table_generator.default_state)
                # Toggle button text.
                if self.toggle_button_span_ctx is not None:
                    with self.toggle_button_span_ctx:
                        if self.table_generator.default_state == _CollapsibleStateImpl.EXPANDED:
                            self.html.addtext("-")
                        else:
                            self.html.addtext("+")

                # Collapsible rows.
                for _collapsible_tr_ctx in self.collapsible_tr_ctxs:  # type: _HtmlDocumentType.NodeContext
                    with _collapsible_tr_ctx:
                        # `.expanded`/`.collapsed` class.
                        self.html.addclass(self.table_generator.default_state)

                        # Visbility.
                        if self.table_generator.default_state == _CollapsibleStateImpl.EXPANDED:
                            self.html.addstyle("visibility: visible")
                        else:
                            self.html.addstyle("visibility: collapse")
